index.rst

```diff
@@ -1,59 +1,367 @@
+Of course, the version in the collection name should differ as little as possible from the versions actually used to reduce confusion.
+
+These names should always correspond to a "public" ``CHAINED`` collection that aggregates both all ``RUN`` collections that directly hold outputs and all collections used as inputs.
+The organization of those "private" output ``RUN`` collections (if there is more than one) is completely at operator discretion, though these collections should start with the same prefix as the umbrella ``CHAINED`` collection, followed by a slash.
+
+In cases where one or more private ``RUN`` collections contain datasets that should not be considered part of the final public outputs (e.g. because they are superceded by datasets in other private ``RUN`` collections), a ``TAGGED`` collection can be used to screen and aggregate these.
+That ``TAGGED`` collection would then be a direct child of the final public ``CHAINED`` collection, instead of any ``RUN`` collections it references.
+
+.. note::
+
+   It is not generally possible to use a ``TAGGED`` collection as the public output collection for a processing run, because putting master calibrations (which are almost always inputs, even if indirectly) in a ``TAGGED`` collection strips them of their validity ranges and does not allow datasets from different validity ranges to coexist.
+   So even if a ``TAGGED`` collection is used, the public ``CHAINED`` collection would contain both that collection and the input ``CALIBRATION`` collection as children.
+
+.. note::
+
+   These public ``CHAINED`` collections essentially mimic Gen2's "parent link" mechanism, which provides at best approximate coarse-grained provenance information about which datasets were used as inputs when producing others.
+   The Gen3 repository will eventually be extended to include fine-grained, exact provenance - essentially a serialization of the directed acyclic graph (DAG) that describes the processing.
+   Whether queries against that DAG are fast enough to allow this more rigorous provenance information to be used as a type of collection (replacing some usage of ``TAGGED`` and ``CHAINED`` collections) remains to be seen, however.
+   It is also worth noting that in general the full DAG does not maintain the usual collection invariant of having only one dataset with a particular dataset type and data ID (e.g. two calexps with the same data ID, from two differently-configured runs, could each contribute to different, non-conflicting coadd patches in downstream runs).
+
+.. _collections-developer-processing-outputs:
+
+Developer processing outputs
+----------------------------
+
+Processing initiated by DM developers that are intended primarily for personal or small-group use must start with ``u/<user>`` (e.g. ``u/jbosch``), and are strongly encouraged to start with ``u/<user>/<ticket>`` (e.g. ``u/jbosch/DM-56789``) whenever possible.
```

Comment:
I think we want to have an explicit note in this section (and/or in "shared outputs" maybe?) that `/u/<username>/*` can CHAIN from anything, but only `/u/<username>/*` collections can be CHAINed to by other `/u/<username>/*` (possibly even requiring that the `<username>` matches?). WIthout such a requirement, if someone uses up their quota and deletes things to clean up, they could mess up other users or "official" collections. I think this should be both a written and in-code-enforced requirement.

Reply:
> I think we want to have an explicit note in this section (and/or in "shared outputs" maybe?) that /u/<username>/* can CHAIN from anything, but only /u/<username>/* collections can be CHAINed to by other /u/<username>/* (possibly even requiring that the <username> matches?).

Agreed, as far as noting that official collections shall not depend on user collections; I'll add a note to that effect.  I do think it's important to allow users to depend on other user's collections in the shared-developer-repo context.  For the Science Platform edition, we'll need more nuance.

Right now, it's actually impossible to delete a collection if it is referenced by a CHAINED collection (including someone else's), so we have some protections in place now.  But you can still delete datasets from that collection, and still break someone else that way.  That's mostly for implementation reasons - it's just a lot harder to prevent.  But hopefully people will tend to delete entire collections rather than individual datasets, and hence it'll be hard to do.

For the curious: dataset deletion is actually one of the hardest things to get right here, because we need to maintain referential integrity between a number of database tables _and_ the content of filesystems, S3, etc. even when somebody else might be trying to write new things with the same data IDs or delete some of the same things concurrently.  A lot of our strategy involves ON DELETE CASCADE foreign key constraints, to shift a lot of that complexity to the database, and so any integrity constraint that _can't_ be captured as a real foreign key in the database is a huge pain.